mono/utils/mono-path.c

```diff
@@ -177,27 +177,137 @@ mono_path_resolve_symlinks (const char *path)
 #endif
 }
 
+static gboolean
+mono_path_char_is_separator (char ch)
+{
+#ifdef HOST_WIN32
+	return ch == '/' || ch == '\\';
+#else
+	return ch == '/';
+#endif
+}
+
+static gboolean
+mono_path_contains_separator (const char *path, size_t length)
+{
+	for (size_t i = 0; i < length; ++i) {
+		if (mono_path_char_is_separator (path [i]))
+			return TRUE;
+	}
+	return FALSE;
+}
+
+static void
+mono_path_remove_trailing_path_separators (const char *path,  size_t *length)
+{
+	size_t i = *length;
+	while (i > 0 && mono_path_char_is_separator (path [i - 1]))
+		i -= 1;
+	*length = i;
+}
+
+#ifdef HOST_WIN32
+
+static gboolean
+mono_path_char_is_lowercase (char ch)
+{
+	return ch >= 'a' && ch <= 'z';
+}
+
+// Version-specific unichar2 upcase tables are stored per-volume at NTFS format-time.
+// This is just a subset.
+static char
+mono_path_char_upcase (char a)
+{
+	return mono_path_char_is_lowercase (a) ? (char)(a - 'a' + 'A') : a;
+}
+
+static gboolean
+mono_path_char_equal (char a, char b)
+{
+	return a == b
+		|| mono_path_char_upcase (a) == mono_path_char_upcase (b)
+		|| (mono_path_char_is_separator (a) && mono_path_char_is_separator (b));
+}
+
+#endif
+
+static gboolean
+mono_path_equal (const char *a, const char *b, size_t length)
+{
+#ifdef HOST_WIN32
+	size_t i = 0;
+	for (i = 0; i < length && mono_path_char_equal (a [i], b [i]); ++i) {
+		// nothing
+	}
+	return i == length;
+#else
+	return memcmp (a, b, length) == 0;
+#endif
+}
+
+static size_t
+mono_path_path_separator_length (const char *a, size_t length)
+{
+	size_t i = 0;
+	while (i < length && mono_path_char_is_separator (a [i]))
+		++i;
+	return i;
+}
+
 /**
  * mono_path_filename_in_basedir:
  *
  * Return \c TRUE if \p filename is in \p basedir
  *
- * Both paths must be absolute and using \c G_DIR_SEPARATOR for directory separators.
+ * Both paths should be absolute and be mostly normalized.
  * If the file is in a subdirectory of \p basedir, returns \c FALSE.
  * This function doesn't touch a filesystem, it looks solely at path names.
+ *
+ * In fact, filename might not be absolute, in which case, FALSE.
+ * Ditto basedir.
```

Comment:
Needs more comments, as indicated later.
Maybe "immediate subdirectory".
/1/2/3 is in /1/2, but /1/2/3/4 is not.
I originally misunderstood the intent.

There are examples below but they aren't as clear as I'd like.
I think I'll do this right now..